mitxgraders/helpers/specify_domain.py

```diff
@@ -0,0 +1,241 @@
+    MathArray([1, 2, 3, 4])
+
+    Provides useful error messages if obj is a number or MathArray:
+    >>> validate_vec4(MathArray([[1, 2, 3], [4, 5, 6]]))
+    Traceback (most recent call last):
+    Invalid: received a matrix of shape (rows: 2, cols: 3), expected a vector of length 4
+    >>> validate_vec4('cat')
+    Traceback (most recent call last):
+    Invalid: received a str, expected a vector of length 4
+    """
+    def shape_validator(obj):
+        if isinstance(obj, MathArray) and obj.shape == shape:
+            return obj
+        raise Invalid("received a {received}, expected a {expected}"
+            .format(received=get_description(obj), expected=MathArray.get_description(shape)))
+
+    return shape_validator
+
+def has_shape(shape):
+    if shape == (1,):
+        return number_validator
+    return make_shape_validator(shape)
+
+# how authors specify domain shape
+SCHEMA_SHAPE = Schema(Any(
+    All(Positive(int), lambda x: tuple([x])),
+    All([Positive(int)], Coerce(tuple)),
+    msg='Expected shape specification to be a positive integer or list of positive integers'
+))
+
+class SpecifyDomain(ObjectWithSchema):
+    """
+    An author-facing, configurable decorator that validates the inputs
+    of the decorated function.
+
+    For now, the only validation available is shape-validation.
+
+    Configuration
+    =============
+    - input_shapes (list): A list of shapes for the function inputs, where:
+        1: means input is scalar
+        k, positive integer: means input is a k-component vector
+        [k1, k2, ...], list of positive integers: means input is an array of shape [k1, k2, ...]
+    - display_name (?str): Function name to be used in error messages
+      defaults to None, meaning that the function's __name__ attribute is used.
+
+    Basic Usage:
+    ============
+
+    Validate that both inputs to a cross product function are 3-component MathArrays:
+    >>> import numpy as np          # just to get the cross product
+    >>> @SpecifyDomain(input_shapes=[3, 3])
+    ... def cross(u, v):
+    ...     return np.cross(u, v)
+
+    If inputs are valid, the function is calculated:
+    >>> a = MathArray([2, -1, 3])
+    >>> b = MathArray([-1, 4, 1])
+    >>> np.array_equal(
+    ...     cross(a, b),
+    ...     np.cross(a, b)
+    ... )
+    True
+
+    If inputs are bad, student-facing DomainErrors are thrown:
+    >>> a = MathArray([2, -1, 3])
+    >>> b = MathArray([-1, 4])
+    >>> cross(a, b)                                 # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+    DomainError: There was an error evaluating function cross(...)
+    ...1st input is ok: received a vector of length 3 as expected
+    ...2nd input has an error: received a vector of length 2, expected a vector of length 3
+    >>> cross(a)
+    Traceback (most recent call last):
+    DomainError: There was an error evaluating function cross(...): expected 2 inputs, but received 1.
+    """
```

Comment:
Can you add a more complicated example here? I'm not sure from the example how to make a function that accepts a scalar and a matrix...

Reply:
In particular, if input_shapes is a list, is that interpreted as a list of shape objects?

Reply:
> In particular, if input_shapes is a list, is that interpreted as a list of shape objects?
Yes, example added.